flush/sink.go

```diff
@@ -0,0 +1,177 @@
+// AckFunc returns an error, it will be returned by `Run`.
+func WithAckFunc(ackFn AckFunc) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackFn = ackFn
+	}
+}
+
+// WithMsgBufferSize sets the msg channel buffer size.
+func WithMsgBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.msgBufferSize = size
+	}
+}
+
+// WithAckBufferSize sets the ack channel buffer size.
+func WithAckBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackBufferSize = size
+	}
+}
+
+// AckFunc is a callback function executed by `Run` for each message succesfully produced. The ack
+// counter is only incremented once this function returns. If no AckFunc is provided the ack counter
+// is always incremented.
+type AckFunc func(msg substrate.Message) error
+
+// AsyncMessageSink wraps substrate.AsyncMessageSink and provides an interface for interaction with
+// the underlying sink, as well as the capability to flush the message buffer.
+type AsyncMessageSink struct {
+	ctx           context.Context
+	cancel        context.CancelFunc
+	wait          func() error
+	sink          substrate.AsyncMessageSink
+	msgs          uint64
+	msgBufferSize int
+	msgCh         chan substrate.Message
+	acks          uint64
+	ackBufferSize int
+	ackCh         chan substrate.Message
+	ackFn         AckFunc
+}
+
+// NewAsyncMessageSink returns a pointer a new AsyncMessageSink.
+// See examples/flush/sink.go for example usage.
+func NewAsyncMessageSink(ctx context.Context, sink substrate.AsyncMessageSink, opts ...AsyncMessageSinkOption) *AsyncMessageSink {
+	ctx, cancel := context.WithCancel(ctx)
+
+	ams := &AsyncMessageSink{
+		ctx:           ctx,
+		cancel:        cancel,
+		sink:          sink,
+		msgBufferSize: defaultMsgBufferSize,
+		ackBufferSize: defaultAckBufferSize,
+	}
+
+	for _, opt := range opts {
+		opt(ams)
+	}
+
+	ams.msgCh = make(chan substrate.Message, ams.msgBufferSize)
+	ams.ackCh = make(chan substrate.Message, ams.ackBufferSize)
+
+	return ams
+}
+
+// Run initialises message publishing using the underlying sink and blocks until either an error
+// occurs or the constructor context is done. If an AckFunc is configured, Run will execute it for
+// each ack recieved. If an error is returned, the user should cancel the constructor context to
+// prevent `Flush` from blocking.
+func (ams *AsyncMessageSink) Run() error {
+	group, groupctx := rungroup.New(ams.ctx)
+	ams.wait = group.Wait
+
+	group.Go(func() error {
+		return ams.sink.PublishMessages(groupctx, ams.ackCh, ams.msgCh)
+	})
+
+	group.Go(func() error {
+		for {
+			select {
+			case <-groupctx.Done():
+				return nil
+			case msg, ok := <-ams.ackCh:
+				if !ok {
+					return nil
+				}
+
+				if ams.ackFn != nil {
+					err := ams.ackFn(msg)
+					if err != nil {
+						return err
+					}
+				}
+
+				atomic.AddUint64(&ams.acks, 1)
+			}
+		}
+	})
```

Comment:
To make sure you process all pending acknowledgments when shutting down, you can do something like this. Though that might slow the shutdown, so depends if we want behaviour like that...
```suggestion
	group.Go(func() error {
		defer close(ams.ackCh)
		return ams.sink.PublishMessages(groupctx, ams.ackCh, ams.msgCh)
	})

	group.Go(func() error {
		for msg := range ams.ackCh {
			if ams.ackFn != nil {
				err := ams.ackFn(msg)
				if err != nil {
					return err
				}
				atomic.AddUint64(&ams.acks, 1)
			}
		}
	})
```